plugins/obs-outputs/ftl-stream.c

```diff
@@ -0,0 +1,1188 @@
+	return pthread_create(&stream->connect_thread, NULL, connect_thread,
+			stream) == 0;
+}
+
+static inline bool add_packet(struct ftl_stream *stream,
+		struct encoder_packet *packet)
+{
+	circlebuf_push_back(&stream->packets, packet,
+			sizeof(struct encoder_packet));
+	return true;
+}
+
+static inline size_t num_buffered_packets(struct ftl_stream *stream)
+{
+	return stream->packets.size / sizeof(struct encoder_packet);
+}
+
+static void drop_frames(struct ftl_stream *stream, const char *name,
+		int highest_priority, bool pframes)
+{
+	UNUSED_PARAMETER(pframes);
+
+	struct circlebuf new_buf            = {0};
+	int              num_frames_dropped = 0;
+
+#ifdef _DEBUG
+	int start_packets = (int)num_buffered_packets(stream);
+#else
+	UNUSED_PARAMETER(name);
+#endif
+
+	circlebuf_reserve(&new_buf, sizeof(struct encoder_packet) * 8);
+
+	while (stream->packets.size) {
+		struct encoder_packet packet;
+		circlebuf_pop_front(&stream->packets, &packet, sizeof(packet));
+
+		/* do not drop audio data or video keyframes */
+		if (packet.type          == OBS_ENCODER_AUDIO ||
+		    packet.drop_priority >= highest_priority) {
+			circlebuf_push_back(&new_buf, &packet, sizeof(packet));
+
+		} else {
+			num_frames_dropped++;
+			obs_encoder_packet_release(&packet);
+		}
+	}
+
+	circlebuf_free(&stream->packets);
+	stream->packets = new_buf;
+
+	if (stream->min_priority < highest_priority)
+		stream->min_priority = highest_priority;
+	if (!num_frames_dropped)
+		return;
+
+	stream->dropped_frames += num_frames_dropped;
+#ifdef _DEBUG
+	debug("Dropped %s, prev packet count: %d, new packet count: %d",
+			name,
+			start_packets,
+			(int)num_buffered_packets(stream));
+#endif
+}
+
+static bool find_first_video_packet(struct ftl_stream *stream,
+		struct encoder_packet *first)
+{
+	size_t count = stream->packets.size / sizeof(*first);
+
+	for (size_t i = 0; i < count; i++) {
+		struct encoder_packet *cur = circlebuf_data(&stream->packets,
+				i * sizeof(*first));
+		if (cur->type == OBS_ENCODER_VIDEO && !cur->keyframe) {
+			*first = *cur;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static void check_to_drop_frames(struct ftl_stream *stream, bool pframes)
+{
+	struct encoder_packet first;
+	int64_t buffer_duration_usec;
+	size_t num_packets = num_buffered_packets(stream);
+	const char *name = pframes ? "p-frames" : "b-frames";
+	int priority = pframes ?
+		OBS_NAL_PRIORITY_HIGHEST : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = pframes ?
+		stream->pframe_drop_threshold_usec :
+		stream->drop_threshold_usec;
+
+	if (num_packets < 5) {
+		if (!pframes)
+			stream->congestion = 0.0f;
+		return;
+	}
+
+	if (!find_first_video_packet(stream, &first))
+		return;
+
+	/* if the amount of time stored in the buffered packets waiting to be
+	 * sent is higher than threshold, drop frames */
+	buffer_duration_usec = stream->last_dts_usec - first.dts_usec;
+
+	if (!pframes) {
+		stream->congestion = (float)buffer_duration_usec /
+			(float)drop_threshold;
+	}
```

Comment:
Style:  Curly braces are not needed for this `if` statement, since the contents of each branch is a single statement.